README.md
# Wagtail Tag Manager

[![CircleCI][circleci]](https://circleci.com/gh/jberghoef/wagtail-tag-manager)
[![TravisCI][travis]](https://travis-ci.org/jberghoef/wagtail-tag-manager)
[![CodeCov][codecov]](https://codecov.io/gh/jberghoef/wagtail-tag-manager)
[![Codacy][codacy]](https://www.codacy.com/app/jberghoef/wagtail-tag-manager)
[![License][license]](https://opensource.org/licenses/BSD-3-Clause)
[![Fossa][fossa]](https://app.fossa.com/projects/git%2Bgithub.com%2Fjberghoef%2Fwagtail-tag-manager)
[![Version][version]](https://pypi.python.org/pypi/wagtail-tag-manager/)
[![Downloads][downloads]](https://pepy.tech/project/wagtail-tag-manager)
[![Black][black]](https://github.com/ambv/black)
[![Prettier][prettier]](https://github.com/prettier/prettier)
[![Gitpod ready-to-code][gitpod]](https://gitpod.io/#https://github.com/jberghoef/wagtail-tag-manager)

[circleci]: https://circleci.com/gh/jberghoef/wagtail-tag-manager.svg?style=svg "Circle CI"
[travis]: https://travis-ci.org/jberghoef/wagtail-tag-manager.svg?branch=master "Travis CI"
[codecov]: https://codecov.io/gh/jberghoef/wagtail-tag-manager/branch/master/graph/badge.svg "Codecov"
[codacy]: https://api.codacy.com/project/badge/Grade/2a59a006e69442bb809bf08f47028eb9 "Codacy"
[license]: https://img.shields.io/badge/license-BSD-blue.svg "License"
[fossa]: https://app.fossa.com/api/projects/git%2Bgithub.com%2Fjberghoef%2Fwagtail-tag-manager.svg?type=shield "Fossa"
[version]: https://img.shields.io/pypi/v/wagtail-tag-manager.svg "Version"
[downloads]: https://pepy.tech/badge/wagtail-tag-manager "Downloads"
[black]: https://img.shields.io/badge/code%20style-black-000000.svg "Black"
[prettier]: https://img.shields.io/badge/code_style-prettier-ff69b4.svg "Prettier"
[gitpod]: https://img.shields.io/badge/Gitpod-ready--to--code-blue?logo=gitpod "Gitpod"

Wagtail Tag Manager (WTM for short) is a [Wagtail](https://wagtail.io/) addon
that allows for easier and GDPR compliant administration of scripts and tags.

**Please note that this software is in beta and has not been extensively tested
in practice. Any part of the API may change without notice!**

![Screenshot](screenshots/screenshot.png "Screenshot")

_In this package the term "tag" is being used for code snippets being injected
into HTML. This is not to be confused with tags used to identify content in the
CMS, such as pictures and documents._

## Features

Wagtail Tag Manager offers functionality similar to platforms like
**Adobe Dynamic Tag Management**, **Google Tag Manager** and **Tealium iQ**
without the need of a third party. It's fully integrated into Wagtail.

This approach comes with a couple of advantages, most prominently the ability
to inject tags into a page before the response is send to a client.

### Administrators

- Manage scripts and tags from within Wagtail, with powerful options to define load order.
- Store reusable content in constants and variables and easily add them to your tags.
- Create triggers to load tags based on events in the browser.
- Create cookie declarations to provide end-users with a full picture of the tracking taking place.

### Developers

- Create custom variables to give administrators functionality specific to your use-cases.

## Table of content

- [Disclaimer](#disclaimer)
- [Requirements](#requirements)
- [Who's using it](#whos-using-it)
- [Instructions](#instructions)
- [Template tags](#template-tags)
  - [`wtm_instant_tags`](#wtm_instant_tags)
  - [`wtm_lazy_manager`](#wtm_lazy_manager)
  - [`wtm_cookie_bar`](#wtm_cookie_bar)
  - [`wtm_include`](#wtm_include)
- [Preference management](#preference-management)
- [Context processors](#context-processors)
- [Settings](#settings)
  - [`WTM_TAG_TYPES`](#wtm_tag_types)
  - [`WTM_INJECT_TAGS`](#wtm_inject_tags)
  - [`WTM_MANAGE_VIEW`](#wtm_manage_view)
  - [`WTM_COOKIE_EXPIRE`](#wtm_cookie_expire)
  - [`WTM_CACHE_TIMEOUT`](#wtm_cache_timeout)
  - [`WTM_PRESERVE_VARIABLES`](#wtm_preserve_variables)
  - [`WTM_INJECT_STYLE`](#wtm_inject_style)
  - [`WTM_INJECT_SCRIPT`](#wtm_inject_script)
  - [`WTM_SUMMARY_PANELS`](#wtm_summary_panels)
  - [`WTM_ENABLE_SCANNER`](#wtm_enable_scanner)
  - [`WTM_CHROMEDRIVER_URL`](#wtm_chromedriver_url)
- [Custom variables](#custom-variables)
- [Page tag mixin](#page-tag-mixin)
- [Lazy triggers](#lazy-triggers)
- [Sandbox](#sandbox)
- [Concept](#concept)
- [License](#license)

## Disclaimer

This package attempts to ease the implementation of tags by the new ePrivacy
rules as defined by the European Union. I urge you to read about these new
rules and ensure you are properly configuring your tags. This package is free
and the author can not be held responsible for the correctness of your
implementation, or the assumptions made in this package to comply with the new
ePrivacy regulation.

Read more about the [ePrivacy Regulation](https://ec.europa.eu/digital-single-market/en/proposal-eprivacy-regulation).

Included in this package is a cookie bar which admittedly provides too little
information to end users regarding the purpose of the scripts you are placing
on the website. For compliance, please use the `cookie_bar.html` template to
change the text shown in the cookie bar.

## Requirements

| Package | Version(s)                                        |
| ------- | ------------------------------------------------- |
| Django  | 2.0, 2.1, 2.2, 3.0, 3.1                           |
| Wagtail | 2.1, 2.2, 2.3, 2.4, 2.5, 2.6, 2.7, 2.8, 2.9, 2.10 |

## Who’s using it?

I'd love to hear from sites and applications where WTM is being used. Please
contact me if you'd like your implementation to be listed here!

## Instructions

**Installation:**

```
pip install wagtail-tag-manager
```

**Add the application to your `INSTALLED_APPS`:**

```python
INSTALLED_APPS = [
    # ...
    'wagtail.contrib.modeladmin',
    'wagtail_tag_manager',
    # ...
]
```

If you wish to enable the cookie bar settings (allowing you to change to title
and text displayed in the cookie bar), also include `wagtail.contrib.settings`
in the `INSTALLED_APPS`.

**Include the middleware:**

```python
MIDDLEWARE = [
    # ...
    "wagtail_tag_manager.middleware.CookieConsentMiddleware",
    'wagtail_tag_manager.middleware.TagManagerMiddleware', # optional
    # ...
]
```

WTM offers two ways to implement it's functionality. You can either choose to
use the `TagManagerMiddleware` (which will rewrite the html on each request)
or use the `{% wtm_instant_tags %}` and `{% wtm_lazy_manager %}` template
tags.

If you prefer to use the template tags to inject tags into your templates,
set the `WTM_INJECT_TAGS` and `WTM_INJECT_SCRIPT` settings to `False`
and implement the template tags as follows:

```html+django
{% load wtm_tags %}

<head>
  {% wtm_instant_tags 'top_head' %} ... {% wtm_instant_tags 'bottom_head' %}
</head>
<body>
  {% wtm_instant_tags 'top_body' %} ... {% wtm_instant_tags 'bottom_body' %} {% wtm_lazy_manager %}
</body>
```

**Include the urls:**

```python
from django.urls import include, path
from wagtail_tag_manager import urls as wtm_urls

urlpatterns = [
    # ...
    path('wtm/', include(wtm_urls)),
    # ...
    path('', include(wagtail_urls)),
    # ...
]
```

## Template tags

As an alternative to using the middleware you can use the `wtm_instant_tags`
and `wtm_lazy_manager` template tags. Please be sure to use the
`TagManagerMiddleware` OR the template tags, never both.

### `wtm_instant_tags`

To load all instant tags at once:

```html+django
{% load wtm_tags %}

<head>
  ... {% wtm_instant_tags %}
</head>
```

To load tags corresponding to a certain position:

```html+django
{% load wtm_tags %}

<head>
  {% wtm_instant_tags 'top_head' %} ... {% wtm_instant_tags 'bottom_head' %}
</head>
<body>
  {% wtm_instant_tags 'top_body' %} ... {% wtm_instant_tags 'bottom_body' %}
</body>
```

### `wtm_lazy_manager`

```html+django
{% load wtm_tags %}

<body>
  ... {% wtm_lazy_manager %}
</body>
```

Optionally, you can disable either the script and/or the styling.

```html+django
{% load wtm_tags %}

<body>
  ... {% wtm_lazy_manager include_style=False include_script=False %}
</body>
```

---

### `wtm_cookie_bar`

![Cookie bar with form](screenshots/cookie-bar-with-form.png "Cookie bar with form")

![Cookie bar with form and details](screenshots/cookie-bar-with-form-and-details.png "Cookie bar with form and details")

```html+django
{% load wtm_tags %}

<body>
  {% wtm_cookie_bar %} ...
</body>
```

### `wtm_include`

WTM comes with the `wtm_include` template tag to accommodate loading of
resources and markup based on the tag strategy and consent given. It can be
used as a way to load html, css or javascript files.

```html+django
{% load wtm_tags %}

<body>
  ... {% wtm_include "necessary" "css/style.css" %} {% wtm_include "necessary" "js/style.js" %} {%
  wtm_include "necessary" "content.html" %} ...
</body>
```

Alternatively, you can use it as a block:

```html+django
{% load wtm_tags %}

<body>
  ... {% wtm_include "statistics" %}
  <script>
    console.log("Included conditionally");
  </script>
  {% wtm_endinclude %} ...
</body>
```

### Preference management

You can use the following provided template tags to render a tag status
overview, a table with cookie declarations or a consent form.

```html+django
{% wtm_tag_table %} {% wtm_declaration_table %} {% wtm_manage_form %}
```

## Context processors

To enable the context processors, add the following to your settings:

```python
"context_processors": [
    # ...
    "wagtail_tag_manager.context_processors.consent_state",
]
```

You can now use the following value in your templates:

```html+django
{{ wtm_consent_state.necessary }} {{ wtm_consent_state.preferences }} {{
wtm_consent_state.statistics }} {{ wtm_consent_state.marketing }}
```

These will return a boolean indicating wether or not tags specific to the
corresponding state should load.

## Setting consent using JavaScript

When building a custom cookie bar or other method of requesting consent, using a
form might not be the most straightforward solution. Use the code example below
to configure cookie consent as an AJAX request.

```javascript
axios({
  method: "POST",
  url: "/wtm/manage/",
  headers: {
    "X-CSRFToken": "[CSRF token]",
    "Content-Type": "application/x-www-form-urlencoded; charset=utf-8",
  },
  data: "necessary=true&preferences=true&statistics=true&marketing=false",
});
```

```javascript
axios({
  method: "POST",
  url: "/wtm/manage/",
  headers: {
    "X-CSRFToken": "[CSRF token]",
    "Content-Type": "application/json; charset=utf-8",
  },
  data: {
    statistics: "true",
    preferences: "true",
    necessary: "true",
    marketing: "false",
  },
});
```

In the resulting response, the correct cookie with the consent state will be
returned.

## Settings

### `WTM_TAG_TYPES`

```python
WTM_TAG_TYPES = {
    # key, verbose name, setting
    "necessary": (_("Necessary"), "required"),
    "preferences": (_("Preferences"), "initial"),
    "statistics": (_("Statistics"), "initial"),
    "marketing": (_("Marketing"), ""),
}
```

Allows you to define the tag types available. This can be helpful if you'd like
the change the terminology used, or when you'd prefer to split a type in
multiple sections. Notice the two keywords (`required` and `initial`) used.

Tags marked as `required` can not be disabled and will always be included on
every page.

Tags marked as `initial` will be included as long as no explicit consent has
been given by the end user, provided the browser allows cookies. While no
consent has been given, these tags will be loaded lazily to honor the browser
settings (which we can only read using javascript).

The third option is to mark a tag as `delayed`. This will ensure the tag will
not load on the first page load, but only from the second load forward.

### `WTM_INJECT_TAGS`

```python
WTM_INJECT_TAGS = True
```

Instructs the middleware to inject all tags marked "instant load" in the
document. Disable this if you would rather use the `{% wtm_instant_tags %}`
template tags.

### `WTM_MANAGE_VIEW`

```python
WTM_MANAGE_VIEW = True
```

Allows you to enable or disable the included "manage" view allowing users to
get insight in the tags running on your site and adjust their preferences.
The view is enabled by default.

### `WTM_COOKIE_EXPIRE`

```python
WTM_COOKIE_EXPIRE = 365
```

Sets the expiration time in days of WTM's cookies. Notice that this is only
applicable to the consent cookies used by WTM, not any cookies placed by tags.

### `WTM_CACHE_TIMEOUT`

```python
WTM_CACHE_TIMEOUT = 1800
```

Sets the amount of seconds the cache will be preserved. At the moment,
caching is only applied to constants, which will refresh when a constant is
saved. Default is 30 minutes.

### `WTM_PRESERVE_VARIABLES`

```python
WTM_PRESERVE_VARIABLES = True
```

Configures whether the variables are preserved for each request, or refreshed
for each tag applied to a response. When set to `False`, a query will be done
for each single tag which will add up quickly.

### `WTM_INJECT_STYLE`

```python
WTM_INJECT_STYLE = True
```

Change to `False` to prevent WTM's included styles from loading. This is useful
if you wish to style the cookiebar yourself.

### `WTM_INJECT_SCRIPT`

```python
WTM_INJECT_SCRIPT = True
```

Change to `False` to prevent WTM's included scripts from loading. This is
useful if you don't want to use the inlcuded lazy loading and cookie bar
functionality.

### `WTM_SUMMARY_PANELS`

```python
WTM_SUMMARY_PANELS = False
```

Disables or enables the summary panels visible on the Wagtail admin dashboard.

![Admin summary panels](screenshots/summary-panels-admin.png "Summary panels on the dashboard")

### `WTM_ENABLE_SCANNER`

**This is an experimental feature.**

```python
WTM_ENABLE_SCANNER = False
```

When enabled, allows scanning of cookies placed on the website. Use this to
automatically generate cookie declarations. Will attempt to use Chrome Driver
when available, and will fall back to regular requests if not.

### `WTM_CHROMEDRIVER_URL`

**This is an experimental feature.**

```python
WTM_CHROMEDRIVER_URL = "http://0.0.0.0:4444/wd/hub"
```

Allows configuration of the docker container running an instance of
`selenium/standalone-chrome`.

When developing, use the following command to run the docker container and
ensure that your site is configured be accessible over your computer's public
ip. Otherwise the docker container won't be able to access the website.

https://hub.docker.com/r/selenium/standalone-chrome/

## Custom variables

In addition to managing variables in the admin interface, variables can also be
created in your source code by registering a `CustomVariable`.

```python
from wagtail_tag_manager.decorators import register_variable
from wagtail_tag_manager.options import CustomVariable

@register_variable
class Variable(CustomVariable):
    name = "Custom variable"
    description = "Returns a custom value."
    key = "custom"

    def get_value(self, request):
        return "This is a custom variable."
```


CHANGES
0.24.2 (2020.10.28)
===================
 - Refactor middleware for added simplicity (#77)
 - Fix CookieConsent location max length (#79)


0.24.1 (2020.10.27)
===================
 - Remove duplicate response functions from middleware


0.24.0 (2020.9.21)
==================
 - Correct admin and trigger tests for missing success messages
 - Add page tags related E2E tests
 - Fully automate dev setup with Gitpod (#76)
 - Fix Cypress integration test for triggers
 - Update front-end dependencies
 - Make cookie scanner functionality optional
 - Update README


0.23.2 (2020.9.15)
==================
 - Update middleware to prevent unnecessary loads
 - Update front-end dependencies


0.23.1 (2020.8.16)
==================
 - Prevent race collisions in middleware (#74)


0.23.0 (2020.8.12)
==================
 - Update README
 - Update front-end dependencies
 - Bump elliptic from 6.5.2 to 6.5.3 (#71)
 - Support for Wagtail 2.10 (#73)
 - Update usage of request.site (#72)


0.22.4 (2020.6.22)
==================
 - Implement end 2 end testing (#69)
 - Create python-publish.yml
 - Update README
 - Dependency updates


0.22.3 (2020.6.18)
==================
 - Dependency updates
 - Add Wagtail 2.9 to testing suite
 - Update example data for Wagtail 2.9


0.22.2 (2020.5.4)
=================
 - Dependency updates


0.22.1 (2020.3.16)
==================
 - Dependency updates


0.22.0 (2020.2.10)
==================
 - Add support for Django 3.0 and Wagtail 2.8 (#66)
 - Upgrade front-end dependencies


0.21.5 (2020.1.29)
==================
 - Split middleware to support non-rewriting implementation


0.21.4 (2020.1.23)
==================
 - Move statics to wagtail_tag_manager folder, fix statics (#65)


0.21.3 (2020.1.15)
==================
 - Fix exampledata


0.21.2 (2020.1.9)
=================
 - Exclude migrations from code quality check
 - Update dependencies


0.21.1 (2019.12.9)
==================
 - Upgrade front-end dependencies


0.21.0 (2019.11.7)
==================
 - Add support for Wagtail 2.7 (#64)


0.20.2 (2019.11.6)
==================
 - Feature/gdpr update (#59)
 - Bump sandbox dependencies
 - Update front-end dependencies
 - Improve Python 3.5 compatability (#60)
 - Add type checking
 - Fix compatibility problem with Sitemap Generator (#62)


0.20.1 (2019.9.2)
=================
 - Optimize the tag manager client
 - Update README to cover lazy triggers
 - Bump lodash from 4.17.11 to 4.17.14 (#58)
 - Update underscore number literals to normal number literals (#57)
 - Update front-end dependencies


0.20.0 (2019.7.5)
=================
 - Add selenium browser tests (#52)
 - Bump fstream from 1.0.11 to 1.0.12 (#53)
 - Add help text blocks (#55)
 - Randomize tests
 - Upgrade front-end dependencies
 - Back-end dependency updates


0.19.1 (2019.5.22)
==================
 - Update and extend README


0.19.0 (2019.5.22)
==================
 - Update README
 - Turn code editor into a widget
 - Reusable widget form horizontal checkbox select multiple
 - Allow the TagMixin to include tags on descending pages
 - Update front-end dependencies
 - Update sandbox requirements


0.18.1 (2019.5.12)
==================
 - Update Twine release script


0.18.0 (2019.5.12)
==================
 - Add included custom variables for the consent state
 - Tweak the tag form view display (#46)
 - Refactor variable system (#47)
 - Lazy triggers and variables (#49)
 - Add Fossa license scans


0.17.1 (2019.4.30)
==================
 - Handle possible None type in cookie request processing


0.17.0 (2019.4.30)
==================
 - Reformat codebase, update Makefile and README
 - Create CODE_OF_CONDUCT.md, update issue templates
 - Restructure models
 - Manage view Django 2.0 compatability
 - Page tag implementation (#32)
 - Switch WebDriver implementation to docker (#45)
 - Add request based fallback to cookie scanner


0.16.6 (2019.4.26)
=================
 - Add content type for long description


0.16.5 (2019.4.26)
==================
 - Add Travis testing Matrix
 - Fix compatibility with Django 2.0
 - Verify compatibility with Django 2.2 and Wagtail 2.5


0.16.4 (2019.4.26)
==================
 - Update front-end bundles


0.16.3 (2019.4.26)
==================
 - Code quality improvements
 - Update README
 - Fix redirect to HomePage stay on current Page (#30) @shaquille-lab
 - Update Codacy configuration
 - Update front-end dependencies


0.16.2 (2019.4.3)
=================
 - Add additional options to lazy manager template tag


0.16.1 (2019.4.3)
=================
 - Always include style and code with lazy manager


0.16.0 (2019.4.3)
=================
 - Repair strategy consent functionality
 - Rename “continue” to “delayed”
 - Add simpler strategy validation (#29)


0.15.1 (2019.4.2)
=================
 - Ensure the lazy manager template tags honors settings


0.15.0 (2019.4.2)
=================
 - Use single cookie for storage of consent state (#28)


0.14.0 (2019.4.1)
=================
 - "Continue" tag type implementation (#25)
 - Custom variable implementation (#26)
 - Rename tag ‘active’ field to ‘auto_load’
 - Allow for filtering of tags in the wtm_instant_tags template tag
 - Make the cookie bar position relative


0.13.2 (2019.2.28)
==================
 - Adjust babel polyfill implementation to avoid collision


0.13.1 (2019.2.25)
==================
 - Fix IE11 list styling error
 - Add babel polyfill for promise support


0.13.0 (2019.2.25)
==================
 - Feature/query optimizations (#24)
 - Split cookie and tag injection logic
 - Tweaks to the trigger form view
 - Add fetch polyfilll for better browser support


0.12.0 (2019.2.19)
==================
 - Simplify cookie bar by using css for tabs (#22)
 - Add FieldPanels to cookie bar settings
 - Front-end dependency updates
 - Update back-end dependencies
 - Remove api consent functionality (#23)


0.11.0 (2019.2.1)
=================
 - Add declaration table to cookie bar
 - Update screenshots and README


0.10.4 (2019.1.31)
==================

src/models.py
import re
import random
import operator

from bs4 import BeautifulSoup
from django.db import models
from django.conf import settings
from django.forms import widgets
from django.dispatch import receiver
from django.template import Context, Template
from django.core.cache import cache
from django.utils.html import mark_safe
from django.core.exceptions import ValidationError
from django.utils.translation import ugettext_lazy as _
from wagtail.admin.edit_handlers import FieldPanel, FieldRowPanel, MultiFieldPanel

from wagtail_tag_manager.managers import (
    TagQuerySet,
    TriggerQuerySet,
    CookieDeclarationQuerySet,
)
from wagtail_tag_manager.settings import TagTypeSettings


class Tag(models.Model):
    TOP_HEAD = "0_top_head"
    BOTTOM_HEAD = "1_bottom_head"
    TOP_BODY = "2_top_body"
    BOTTOM_BODY = "3_bottom_body"
    LOCATION_CHOICES = [
        (TOP_HEAD, _("Top of head tag")),
        (BOTTOM_HEAD, _("Bottom of head tag")),
        (TOP_BODY, _("Top of body tag")),
        (BOTTOM_BODY, _("Bottom of body tag")),
    ]

    INSTANT_LOAD = "instant_load"
    LAZY_LOAD = "lazy_load"
    LOAD_CHOICES = [(INSTANT_LOAD, _("Instant")), (LAZY_LOAD, _("Lazy"))]

    name = models.CharField(max_length=100, unique=True)
    description = models.TextField(null=True, blank=True)
    active = models.BooleanField(
        default=True,
        help_text=_(
            "Uncheck to disable this tag from being included, "
            "or when using a trigger to include this tag."
        ),
    )
    priority = models.SmallIntegerField(
        default=0,
        help_text=mark_safe(
            _(
                "Define how early on this tag should load as compared to other tags. "
                "A higher number will load sooner. For example:<br/>"
                " - A tag with a priority of 3 will load before a tag with priority 1.<br/>"
                " - A tag with a priority 0 will load before a tag with priority -1.<br/>"
                "<em>Please note that with instanly loading tags, the priority is only "
                "compared to tags that load in the same document location.</em>"
            )
        ),
    )

    tag_type = models.CharField(
        max_length=10,
        choices=[(key, _(key.title())) for key in TagTypeSettings.all().keys()],
        default=list(TagTypeSettings.all())[0],
        help_text=_(
            "The purpose of this tag. Will decide if and when this tag is "
            "loaded on a per-user basis."
        ),
    )
    tag_location = models.CharField(
        max_length=14,
        choices=LOCATION_CHOICES,
        default=TOP_HEAD,
        help_text=_(
            "Where in the document this tag will be inserted. Only applicable "
            "for tags that load instantly."
        ),
    )
    tag_loading = models.CharField(
        max_length=12,
        choices=LOAD_CHOICES,
        default=INSTANT_LOAD,
        help_text=mark_safe(
            _(
                "<b>Instant:</b> include this tag in the document when the initial "
                "request is made.<br/>"
                "<b>Lazy:</b> include this tag after the page has finished loading."
            )
        ),
    )

    content = models.TextField(
        help_text=_(
            "The tag to be added or script to be executed."
            "Will assume the content is a script if no explicit tag has been added."
        )
    )

    objects = TagQuerySet.as_manager()

    panels = [
        FieldPanel("name", classname="full title"),
        FieldPanel("description", classname="full"),
        MultiFieldPanel(
            [
                FieldPanel("tag_type"),
                FieldRowPanel([FieldPanel("tag_loading"), FieldPanel("tag_location")]),
                FieldPanel("priority"),
                FieldPanel("active"),
            ],
            heading=_("Meta"),
            classname="collapsible",
        ),
        FieldPanel("content", classname="full code"),
    ]

    class Meta:
        ordering = ["tag_loading", "-active", "tag_location", "-priority"]

    def clean(self):
        if not re.match("\<.+\/?\>", self.content):
            self.content = f"<script>{self.content}</script>"

        self.content = BeautifulSoup(self.content, "html.parser").prettify()

        return self

    def save(
        self, force_insert=False, force_update=False, using=None, update_fields=None
    ):
        self.full_clean()
        return super().save(force_insert, force_update, using, update_fields)

    def get_doc(self, request=None, context=None):
        content = self.content

        if request:
            template = Template(content)
            context = Context(Tag.create_context(request, context))
            content = template.render(context)

        return BeautifulSoup(content, "html.parser")

    @classmethod
    def create_context(cls, request, context=None):
        context = context or {}
        if hasattr(context, "flatten"):
            context = context.flatten()

        return {
            **Constant.create_context(),
            **Variable.create_context(request),
            **context,
        }

    @classmethod
    def get_types(cls):
        return list(TagTypeSettings.all())

    @classmethod
    def get_cookie_name(cls, tag_type):
        return f"wtm_{tag_type}"

    def __str__(self):
        return self.name


class Constant(models.Model):
    name = models.CharField(max_length=100, unique=True)
    description = models.TextField(null=True, blank=True)

    key = models.SlugField(
        max_length=255,
        unique=True,
        help_text=mark_safe(
            _(
                "The key that can be used in tags to include the value.<br/>"
                "For example: <code>{{ ga_id }}</code>."
            )
        ),
    )
    value = models.CharField(
        max_length=255,
        help_text=_("The value to be rendered when this constant is included."),
    )

    panels = [
        FieldPanel("name", classname="full title"),
        FieldPanel("description", classname="full"),
        MultiFieldPanel(
            [FieldRowPanel([FieldPanel("key"), FieldPanel("value")])], heading=_("Data")
        ),
    ]

    def as_dict(self):
        return {
            "name": self.name,
            "description": self.description,
            "key": self.key,
            "value": self.value,
        }

    def get_value(self):
        return self.value

    @classmethod
    def create_context(cls):
        context = cache.get("wtm_constant_cache", {})

        if not context:
            for constant in cls.objects.all():
                context[constant.key] = constant.get_value()

            timeout = getattr(settings, "WTM_CACHE_TIMEOUT", 60 * 30)
            cache.set("wtm_constant_cache", context, timeout)

        return context

    def clean(self):
        if Variable.objects.filter(key=self.key).exists():
            raise ValidationError(
                f"A variable with the key '{ self.key }' already exists."
            )
        else:
            super().clean()

    def save(
        self, force_insert=False, force_update=False, using=None, update_fields=None
    ):
        self.full_clean()
        return super().save(force_insert, force_update, using, update_fields)

    def __str__(self):
        return self.name


@receiver(models.signals.post_save, sender=Constant)
def handle_constant_save(sender, **kwargs):
    sender.create_context()  # Update the cache


class Variable(models.Model):
    TYPE_CHOICES = (
        (_("HTTP"), (("path", _("Path")), ("_repath+", _("Path with regex")))),
        (_("User"), (("user.pk", _("User")), ("session.session_key", _("Session")))),
        (_("Wagtail"), (("site", _("Site")),)),
        (_("Other"), (("_cookie+", _("Cookie")), ("_random", _("Random number")))),
    )

    name = models.CharField(max_length=100, unique=True)
    description = models.TextField(null=True, blank=True)

    key = models.SlugField(
        max_length=255,
        unique=True,
        help_text=mark_safe(
            _(
                "The key that can be used in tags to include the value.<br/>"
                "For example: <code>{{ path }}</code>."
            )
        ),
    )
    variable_type = models.CharField(
        max_length=255,
        choices=TYPE_CHOICES,
        help_text=mark_safe(
            _(
                "<b>Path:</b> the path of the visited page.<br/>"
                "<b>Path with regex:</b> the path of the visited page after "
                "applying a regex search.<br/>"
                "<b>User:</b> the ID of a user, when available.<br/>"
                "<b>Session:</b> the session key.<br/>"
                "<b>Site:</b> the name of the site.<br/>"
                "<b>Cookie:</b> the value of a cookie, when available.<br/>"
                "<b>Random number:</b> a random number."
            )
        ),
    )
    value = models.CharField(
        max_length=255,
        null=True,
        blank=True,
        help_text=mark_safe(
            _(
                "<b>Path with regex:</b> the pattern to search the path with.<br/>"
                "<b>Cookie:</b> the name of the cookie."
            )
        ),
    )

    panels = [
        FieldPanel("name", classname="full title"),
        FieldPanel("description", classname="full"),
        MultiFieldPanel(
            [
                FieldRowPanel([FieldPanel("key"), FieldPanel("variable_type")]),
                FieldPanel("value"),
            ],
            heading=_("Data"),
        ),
    ]

    def as_dict(self):
        return {
            "name": self.name,
            "description": self.description,
            "key": self.key,
            "variable_type": self.variable_type,
            "value": self.value,
        }

    def get_repath(self, request):
        path = request.path
        if self.value:
            regex = re.compile(self.value)
            match = regex.search(request.get_full_path())
            if match:
                return match.group()
            return ""
        return path

    def get_cookie(self, request):
        return request.COOKIES.get(self.value, "")

    def get_random(self, request):
        return int(random.random() * 2_147_483_647)

    def get_value(self, request):
        variable_type = self.variable_type

        if variable_type.endswith("+"):
            variable_type = variable_type[:-1]

        if variable_type.startswith("_"):
            method = getattr(self, f"get{variable_type}")
            return method(request)

        if "." in self.variable_type:
            return operator.attrgetter(str(self.variable_type))(request)

        return getattr(request, str(self.variable_type))

    @classmethod
    def create_context(cls, request):
        context = {}

        for variable in cls.objects.all():
            context[variable.key] = variable.get_value(request)

        return context

    def clean(self):
        if Constant.objects.filter(key=self.key).exists():
            raise ValidationError(
                f"A constant with the key '{ self.key }' already exists."
            )
        else:
            super().clean()

            if not self.variable_type.endswith("+"):
                self.value = ""

            return self

    def save(
        self, force_insert=False, force_update=False, using=None, update_fields=None
    ):
        self.full_clean()
        return super().save(force_insert, force_update, using, update_fields)

    def __str__(self):
        return self.name

def searchable_regex_validator(value):
    try:
        re.search(value, "")
    except re.error:
        raise ValidationError("The pattern {pattern} is not valid".format(pattern=value))

class Trigger(models.Model):
    name = models.CharField(max_length=100, unique=True)
    description = models.TextField(null=True, blank=True)

    active = models.BooleanField(
        default=True, help_text=_("Uncheck to disable this trigger from firing.")
    )
    pattern = models.CharField(
        max_length=255,
        help_text=_(
            "The regex pattern to match the full url path with. "
            "Groups will be added to the included tag's context."
        ),
        validators=[searchable_regex_validator]
    )
    tags = models.ManyToManyField(
        Tag, help_text=_("The tags to include when this trigger is fired.")
    )

    objects = TriggerQuerySet.as_manager()

    panels = [
        FieldPanel("name", classname="full title"),
        FieldPanel("description", classname="full"),
        MultiFieldPanel(
            [FieldPanel("pattern"), FieldPanel("active")], heading=_("Configuration")
        ),
        FieldPanel("tags", widget=widgets.CheckboxSelectMultiple),
    ]

    def match(self, request):
        return re.search(self.pattern, request.get_full_path())

    def __str__(self):
        return self.name


class CookieDeclaration(models.Model):
    INSECURE_COOKIE = "http"
    SECURE_COOKIE = "https"
    SECURITY_CHOICES = ((INSECURE_COOKIE, _("HTTP")), (SECURE_COOKIE, _("HTTPS")))

    cookie_type = models.CharField(
        max_length=10,
        choices=[(key, _(key.title())) for key in TagTypeSettings.all().keys()],
        help_text=_("The type of functionality this cookie supports."),
        null=True,
        blank=True,
    )
    name = models.CharField(max_length=255, help_text=_("The name of this cookie."))
    domain = models.CharField(
        max_length=255,
        help_text=mark_safe(
            _(
                "The domain (including subdomain if applicable) of the cookie.<br/>"
                "For example: <code>.wagtail.io</code>."
            )
        ),
    )
    purpose = models.TextField(help_text=_("What this cookie is being used for."))
    duration = models.DurationField(null=True, blank=True)
    security = models.CharField(
        max_length=5,
        choices=SECURITY_CHOICES,
        default=INSECURE_COOKIE,
        help_text=_("Whether this cookie is secure or not."),
    )

    objects = CookieDeclarationQuerySet.as_manager()

    panels = [
        FieldPanel("name", classname="full title"),
        MultiFieldPanel(
            [
                FieldPanel("cookie_type"),
                FieldPanel("purpose"),
                FieldPanel("duration"),
                FieldRowPanel([FieldPanel("domain"), FieldPanel("security")]),
            ],
            heading=_("General"),
        ),
    ]

    class Meta:
        ordering = ["domain", "cookie_type", "name"]
        unique_together = ("name", "domain")

    def save(
        self, force_insert=False, force_update=False, using=None, update_fields=None
    ):
        self.full_clean()
        return super().save(force_insert, force_update, using, update_fields)

    @property
    def expiration(self):
        return self.duration

    def __str__(self):
        return self.name
